presto-delta/src/main/java/com/facebook/presto/delta/DeltaPageSourceProvider.java

```diff
@@ -0,0 +1,402 @@
+    {
+        if (effectivePredicate.isNone()) {
+            return TupleDomain.none();
+        }
+
+        ImmutableMap.Builder<ColumnDescriptor, Domain> predicate = ImmutableMap.builder();
+        for (Map.Entry<DeltaColumnHandle, Domain> entry : effectivePredicate.getDomains().get().entrySet()) {
+            DeltaColumnHandle columnHandle = entry.getKey();
+
+            RichColumnDescriptor descriptor;
+
+            if (isPushedDownSubfield(columnHandle)) {
+                Subfield pushedDownSubfield = getPushedDownSubfield(columnHandle);
+                List<String> subfieldPath = columnPathFromSubfield(pushedDownSubfield);
+                descriptor = descriptorsByPath.get(subfieldPath);
+            }
+            else {
+                descriptor = descriptorsByPath.get(ImmutableList.of(columnHandle.getName()));
+            }
+
+            if (descriptor != null) {
+                predicate.put(descriptor, entry.getValue());
+            }
+        }
+        return TupleDomain.withColumnDomains(predicate.build());
+    }
+
+    public static Optional<org.apache.parquet.schema.Type> getParquetType(
+            Type prestoType,
+            MessageType messageType,
+            DeltaColumnHandle column,
+            SchemaTableName tableName,
+            Path path)
+    {
+        org.apache.parquet.schema.Type type = getParquetTypeByName(column.getName(), messageType);
+        if (type == null) {
+            return Optional.empty();
+        }
+
+        if (!checkSchemaMatch(type, prestoType)) {
+            String parquetTypeName;
+            if (type.isPrimitive()) {
+                parquetTypeName = type.asPrimitiveType().getPrimitiveTypeName().toString();
+            }
+            else {
+                GroupType group = type.asGroupType();
+                StringBuilder builder = new StringBuilder();
+                group.writeToStringBuilder(builder, "");
+                parquetTypeName = builder.toString();
+            }
+            throw new PrestoException(
+                    DELTA_PARQUET_SCHEMA_MISMATCH,
+                    format("The column %s of table %s is declared as type %s, but the Parquet file (%s) declares the column as type %s",
+                            column.getName(),
+                            tableName.toString(),
+                            column.getDataType(),
+                            path.toString(),
+                            parquetTypeName));
+        }
+        return Optional.of(type);
+    }
+
+    public static boolean isPushedDownSubfield(DeltaColumnHandle column)
+    {
+        return column.getColumnType() == SUBFIELD;
+    }
+
+    /**
+     * Return the pushed down subfield if the column represents one
+     */
+    public static Subfield getPushedDownSubfield(DeltaColumnHandle column)
+    {
+        checkArgument(isPushedDownSubfield(column), format("not a valid pushed down subfield: %s", column));
+        return column.getSubfield().get();
```

Comment:
`column.getSubfield().get()` is called without a isPresent() check. Please add the validation.

Reply:
Its part of the `isPushedDownSubfield` check above.